ipywidgets/widgets/widget.py

```diff
@@ -50,6 +50,47 @@ def _json_to_widget(x, obj):
 else:
     _binary_types = (memoryview, buffer)
 
+
+def _split_state_buffers(state):
+    """Return (state_without_buffers, buffer_paths, buffers) for binary message parts
+
+    As an example:
+    >>> state = {'plain': [0, 'text'], 'x': {'ar': memoryview(ar1)}, 'y': {'shape': (10,10), 'data': memoryview(ar2)}}
+    >>> widget._split_state_buffers(state)
+    ({'plain': [0, 'text']}, {'x': {}, 'y': {'shape': (10, 10)}}, [['x', 'ar'], ['y', 'data']],
+     [<memory at 0x107ffec48>, <memory at 0x107ffed08>])
+    """
+
+    def seperate_buffers(substate, path, buffer_paths, buffers):
+        # remove binary types from dicts and lists, but keep track of their paths
+        # e.g. {'x': {'ar': ar}, 'y': [ar2, ar3]}, where ar/ar2/ar3 are binary types
+        # will result in {'x': {}, 'y': [None, None]}, [ar, ar2, ar3], [['x', 'ar'], ['y', 0], ['y', 1]]
+        # instead of removing elements from the list, this will make replacing the buffers on the js side much easier
+        if isinstance(substate, (list, tuple)):
+            for i, v in enumerate(substate):
+                if isinstance(v, (dict, list, tuple)):
+                    seperate_buffers(v, path + [i], buffer_paths, buffers)
+                if isinstance(v, _binary_types):
+                    substate[i] = None
+                    buffers.append(v)
+                    buffer_paths.append(path + [i])
+        elif isinstance(substate, dict):
+            for k, v in list(substate.items()):  # we need to copy to a list since substate will be modified
```

Comment:
This code will not be the bottleneck, and I think that deleting afterwards makes the code more difficults to read, let me know what you think of the (changed) code.